r/R/alchemize.R

```diff
@@ -0,0 +1,81 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#' Transform a data structure from one engine to another
```

Comment:
I see what you're trying to do here but I'm not convinced that we need it (yet, at least). 

What if instead we defined `tbl.arrow_dplyr_query` et al.-- it is a generic in `dplyr`--so the tests would look like:

```r
ds %>%
  select(int, lgl) %>%
  tbl(.engine = "duckdb") %>%
  summarize(...)
```

Reply:
Oh, yeah that's not too far off of what `tbl()` is used for elsewhere both `collect()` and `compute()` imply that there's a data processing/change step, where as `tbl()` doesn't have that same baggage. `tbl()` is a bit funny as it stands: [with `tbl.src_dbi()`](https://github.com/tidyverse/dbplyr/blob/47e53ce30402d41ae4b38c803de12e63d64a9b6c/R/src_dbi.R#L83-L86) takes a `src` (connection), and `from` a character of the table name versus the standard `tbl` from dplyr where the `src` argument is the thing you want to create a `tbl` from. Which would make our `tbl.*()` funny in that it has arguments: `src` - the Arrow object, `con` - the db connection to use, and `name` (or something like it, or maybe we embed that inside the function...) – the name of the view to create. 

I think for now this is fine (though I will try and make sure that the docs explain the funny bit where _for arrow -> duckdb_ the `src` in `tbl` is *not* the duckdb dbi connection like one might expect (and instead there's a con argument for that)

Reply:
🤔 I'm working on this right now and I'm torn. I like the idea of having one function / family of them that clearly do this kind of transformation (right now it's relatively limited, but we've got 2 already!).

`tbl` works pretty well, though right now I'm trying to figure out what the right answer is for `tbl(..., .engine = "arrow")`. In an ideal world(?) duckdb would also expose a C-interface for us to get Arrow data out of, such that one could pull data back in to Arrow without needing to route via R which this would be exactly that (though AFAICT that doesn't currently exist).

I have right now (with the comment in the middle not totally sure what to do)
```
tbl.arrow_dplyr_query <- function(src, ..., .engine = c("arrow", "duckdb")) {
  engine <- match.arg(.engine)

  if (engine == "arrow") {
    # What should we do here? Possibilities:
    # stop("not yet implemented")?
    # return(src) # no-op + maybe a warning that we did nothing
    # return(arrow_dplyr_query(src)) # this is the most tbl like thing for an Arrow object
  } else if (engine == "duckdb") {
    .alchemize_to_duckdb(src = src, ...)
  }
}
```

So maybe it is worth waiting until some of those other features exist (or the next hand-over integration comes up) before we come up with a grand scheme that unifies these kinds of hand offs.